Use a standard IV length for AES keys

sfackler opened the pull request:
We still support decryption of values with the old IV size, but newly
encrypted values use the proper length.

This assumes that we don't expect old versions of this library to be decrypting values produced by a newer version of it. Is that correct to think?

Closes #55

markelliot commented:
@gdearment does this affect any of your stuff?

bmoylan commented:
Yeah -- the the golang implementation (currently internal) will need to be updated as well. That is what our deployment tooling uses to generate the encrypted values (not this java impl) so we'll need to add compatibility in go, get all products sporting the new dependency, then update the config generation code to use the new length once we're confident everything supports it.

markelliot commented:
So seems like the real change here should be emitting the IV length into the file format somehow, and if not finding it, then falling back, rather than changing the hard-coded thing.

gdearment commented:
Yea, this has come up on some of the Apollo infrastructure stuff where we use AES encryption into kinesis. It would be nice to encode information about the key into the key, similar to how i believe it was done for hadoop-crypto.

markelliot commented:
Ok, so I think the solution, then, is to change the storage format, try to parse as the new format and then fallback to the old default.

Then, changing what the internal stuff produces is going to be a bit of a pain, but maybe we can play games with our application manifests to hint which style to produce?

sfackler commented:
Adding a tag to the value format makes sense to me. I guess it'd probably make sense to mirror what we do with keys. We could use just `AES` for the tag like we do for keys, but it might be better to use something a bit more specific like `GCM`, `AES-GCM` or even `AES-256-GCM`?

sfackler commented:
Ok, I've added a commit that adds an algorithm prefix to encrypted values, and the AES side uses the existence of that to know which IV size to use. It is backwards compatible in that it can read old-style values, but it will always produce new-style values.

bavardage commented:
@markelliot does the above seem ok?

markelliot commented on encrypted-config-value/src/main/java/com/palantir/config/crypto/EncryptedValue.java:
```diff
@@ -18,21 +18,32 @@
 
-    public abstract String encryptedValue();
+    @Nullable
+    public abstract String getAlgorithm();
```
would prefer Optional vs. Nullable

markelliot replied:
though in reality I think we should always populate this field, when it's not found during parsing we can put the default value into it, which makes the remaining consumer code not have to branch

sfackler replied:
Fixed

sfackler replied:
Didn't see the second comment - we could fill in with `UNKNOWN` or something like that? We'd probably have to check for that in the `toString` impl so that old-style values round-trip through `EncryptedValue` without gaining a prefix.

sfackler replied:
Oh, actually, consumer code has to branch either way since we need to support prefix-less values.

markelliot replied:
I guess it's strange to me that we're not encoding the actual thing we're switching on in the format and decoding that into this class.

sfackler replied:
We are - encrypted values produced after this change look like `${enc:AES:3Jw0mA0TbPOqpqskhMx2hqC5cdfTWG/nPi0ut36cPkPax5ZL1uWrVJOVhTZ/LfbsO1H6ZLVXP0b/Ng3quYvvYnK2Sx9+KZM=}`. The `AES` bit in there is parsed and stuck in the algorithm field of `EncryptedValue`. The issue is that we still need to support the old form (e.g. `${enc:QjR4AHIYoIzvjEHf53XETM3QYnCl1mgFYC51Q7x4ebwM+h3PHVqSt/1un/+KvpJ2mZfMH0tifu+htRVxEPyXmt88lyKB83NpesNJEoLFLL+wBWCkppaLRuc/1w==}`) which is missing the algorithm specifier. Upstream deployment infrastructure stuff will still be producing this old format for some amount of time since it seems unnecessary to break deserialization back compat and force the entire ecosystem to upgrade at once. When we do decide to make that break, we can remove the `Optional` here.

markelliot replied:
"AES" doesn't sufficiently encode the parameter space. The current algorithm is AES, but has a different IV length. Ideally we'd also be encoding the IV length, and we'd set the default for these missing parameters to be AES and the current IV length.

sfackler replied:
I picked "AES" since it was what we were tagging the keys with - "AES-256-GCM" would be a full description if you'd prefer to switch to that. I don't think it's necessary to explicitly state the IV size since there is a single canonical value of 92 bits. The IV size cannot be changed at all in most AES cipher modes. Even though it can with GCM, it's not a thing that happens very commonly (and is actively discouraged as #55 notes).

markelliot commented:
Yet explicitly this is something that was however painfully or uncommonly configured. Ideally we're not special casing one variant of AES-256-GCM against another by absence of information. Either it's fully encoded in an algorithm specific way or we use the present default. I don't think it's reasonable to label the "new" formats as AES or even AES-256-GCM because the old formats would logically also be just fine to receive this label.

On Mar 20, 2017, at 10:35 PM, Steven Fackler <notifications@github.com<mailto:notifications@github.com>> wrote:


@sfackler commented on this pull request.

________________________________

In encrypted-config-value/src/main/java/com/palantir/config/crypto/EncryptedValue.java[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_palantir_encrypted-2Dconfig-2Dvalue_pull_56-23discussion-5Fr107033882&d=DwMCaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=Vp81aOxWZvuVgVK_wp-VF3pIYG92B19LcCw6XKeYC0U&m=y49c3jqnEJaC_BZTP3vyDLmSqz7xa82ZxhKzpqOwmeM&s=qTiMh9kHC6Xldbtk8TVtFnck25isIrmfwra0UcctjhM&e=>:

>  import org.immutables.value.Value;

 @Value.Immutable
 public abstract class EncryptedValue {

     private static final String PREFIX = "enc:";

-    public abstract String encryptedValue();
+    @Nullable
+    public abstract String getAlgorithm();


I picked "AES" since it was what we were tagging the keys with - "AES-256-GCM" would be a full description if you'd prefer to switch to that. I don't think it's necessary to explicitly state the IV size since there is a single canonical value of 92 bits. The IV size cannot be changed at all in most AES cipher modes. Even though it can with GCM, it's not a thing that happens very commonly (and is actively discouraged as #55[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_palantir_encrypted-2Dconfig-2Dvalue_issues_55&d=DwMCaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=Vp81aOxWZvuVgVK_wp-VF3pIYG92B19LcCw6XKeYC0U&m=y49c3jqnEJaC_BZTP3vyDLmSqz7xa82ZxhKzpqOwmeM&s=GYrXdf-qqVipuYm-L95wA6wBZLWCKKQTgndAwuyzYpY&e=> notes).

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_palantir_encrypted-2Dconfig-2Dvalue_pull_56-23discussion-5Fr107033882&d=DwMCaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=Vp81aOxWZvuVgVK_wp-VF3pIYG92B19LcCw6XKeYC0U&m=y49c3jqnEJaC_BZTP3vyDLmSqz7xa82ZxhKzpqOwmeM&s=qTiMh9kHC6Xldbtk8TVtFnck25isIrmfwra0UcctjhM&e=>, or mute the thread[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAp3Q1m8dFGd9qTIUZVor3UA-2Dxae7Esgks5rnv81gaJpZM4MN3aA&d=DwMCaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=Vp81aOxWZvuVgVK_wp-VF3pIYG92B19LcCw6XKeYC0U&m=y49c3jqnEJaC_BZTP3vyDLmSqz7xa82ZxhKzpqOwmeM&s=28U3QlEdLeZ12lCQ7EiTFUvhvpyrBmIk4W7pgbb16qQ&e=>.

sfackler commented:
I think a reasonable thing to do there would be so split the encrypted value portion into a set of colon-separated portions. RSA would just have a single part, but the AES side could explicitly split out the IV, ciphertext, and tag. One thing I would want to make explicit here is that the only valid configuration (other than the old legacy setup) is a 92 bit IV and a 128 bit tag. The format may syntactically support e.g. a 256 bit IV and 32 bit tag, but that's not something that should actually be created anywhere.

@akerl was advocating a switch over to using something more predetermined like [crypto_secretbox](https://download.libsodium.org/doc/secret-key_cryptography/authenticated_encryption.html) for exactly these reasons. That seems like a good idea to me though I'm not sure if there are great Java bindings available. The one I've found that looks reasonably popular appears to want to link to the system libsodium instead of vendoring a shared library.

markelliot commented:
sure...so the original suggestion we'd had was to duplicate https://github.com/palantir/hadoop-crypto/blob/develop/crypto-serialization/src/main/java/com/palantir/crypto2/KeySerializerV2.java.

sfackler commented:
Ah, sorry, must have missed that. We could use a binary encoding like hadoop-crypto, but it seems like it would just add a bit of unnecessary complication on the encode and decode sides - the encrypted value has to end up as a string in the end in this case, and `s.split(':')` is  simpler than having to  read length-prefixed entries out manually.

nmiyake commented:
Agree that keeping string format would be nice... Seems like we would need to either go down the path of encoding all of this information as part of the "name" of the algorithm, or add extra information.

So something like: `enc:AES-256-GCM-12:<content>`

The alternative would be to have the things like key length and field each be separate fields:

`enc:AES-GCM:256:12:<content>`

However, issue there is that not all fields are relevant for all keys -- RSA keys don't have an IV/nonce length, so then we would end up with something like:

`enc:RSA:1024::<content>` or `enc:RSA:1024:0:<content>`

which may be confusing to read through.

sfackler commented:
The approach I mentioned above would look something like `enc:AES-256-GCM;<iv>:<ciphertext>:<tag>`, or `enc:RSA:<ciphertext>`. Since we're identifying the algorithm up front, the rest of the string can be algorithm-specific.

nmiyake commented:
OK yup good point, agree that as long as algorithm comes first the rest can be variable, which is nice.

Although I understand the point that Mark is making in https://github.com/palantir/encrypted-config-value/pull/56#issuecomment-287921141, I'm partial to your view that explicitly specifying the IV and tag length when the two should basically never be configurable seems a bit strange -- it adds cognitive overhead (specifying the values makes it seem like something that can be configured), and the only reason that's cited is because we chose bad (non-standard) defaults in the first implementation.

Even if we decide that we have to make the IV length part of the encoded specification, I would advocate for not including the tag length in the new format -- we never expect it to be configured, so don't think it's worth the extra effort/confusion it may introduce.

sfackler commented:
The tag size is like the IV size with AES GCM - it can be selected (I think it just truncates the full tag down, but it is in the standard), but you should only ever be using the default. It seems to me like we should either force the standard sizes for both or explicitly encode the sizes for both.

nmiyake commented:
Conceptually I definitely agree, and my preference would be "force standard size for both, except for when we have to fall back to the old format". However, understand Mark's concern that this could be confusing.

For the IV, we already have a non-standard size floating around, so that's a real thing to consider. I'm saying that, even though it's technically less consistent to expose only IV size and not tag size, I think the benefit of not having another thing that could end up becoming non-standard outweighs the benefits of consistency ("we expose IV size, so we should expose tag size").

Not dead-set on this, but those are my instincts.

nmiyake commented:
Concrete Proposal
-----------------
* Key format should stay mostly the same, with the only modification being that the "algorithm" portion should uniquely identify how the content should be read (currently, "RSA:" may indicate a public or private key). 

General form: `<key-alg-name>:{b64-encoded-key-material}`

Examples:
  * `AES:<b64-encoded-key-bytes>`
  * `RSA-PUB:<b64-encoded-PEM-public-key>`
  * `RSA-PRIV:<b64-encoded-DER-PKCS8-private-key>`

* Encrypted values **produced** by updated library should include encryption algorithm and parameters used by the algorithm to produce the encrypted value.

General form: `${enc:<key-alg-name>:<b64-encoded-ciphertext>(:<param>)*}`.

Examples:
  * `${enc:AES:<b64-encoded-ciphertext>:GCM:<iv-len-bytes>:<tag-len-bytes>}`
  * `${enc:RSA-PUB:<b64-encoded-ciphertext>}`

Background
------------
Chatted through this with @sfackler in person today and here's what we came up with (there are also a few more modification I included, so not implying that Steven is on board with everything):

* High-level overview:
  * A key should encode only the information needed to construct the key itself
  * An encrypted value should contain information about the key type and the algorithm and parameters used to generate the data

* Key format should stay mostly the same, with the only modification being that the "algorithm" portion should uniquely identify how the content should be read (currently, "RSA:" may indicate a public or private key). The general form is `<key-alg-name>:{b64-encoded-key-material}`.
  * `AES:<b64-encoded-key-bytes>`
  * `RSA-PUB:<b64-encoded-PEM-public-key>`
  * `RSA-PRIV:<b64-encoded-DER-PKCS8-private-key>`

* Encrypted values **produced** by updated library will include encryption algorithm and parameters used by the algorithm to produce the encrypted value. The general form is `${enc:<key-alg-name>:<b64-encoded-ciphertext>(:<param>)*}`.
  * `${enc:AES:<b64-encoded-ciphertext>:GCM:<iv-len-bytes>:<tag-len-bytes>}`
  * `${enc:RSA-PUB:<b64-encoded-ciphertext>}`

The components are delimited by `:` and parsing can be conditional based on values encounterd so far. For example, it could be the case that `enc:AES:<b64>:GCM` -> `:<iv-len-bytes>:<tag-len-bytes>` required, while `enc:AES:<b64>:OCB` -> different set of parameters required.

For the AES-GCM encryption performed by the encrypted-config-value library, the values produced by the updated library should always have a `GCM` tag, `iv-len-bytes` of 12 and `tag-len-bytes` of 16, so values will be of the form `${enc:AES:<b64-encoded-ciphertext>:GCM:12:16`.

General motivation is that, given an encrypted-value, it should be possible to know how it was produced (which will also inform how it should be decrypted). The scheme proposed above also maintains a mapping between the `<key-alg-name>` used in the key's storage format and the encrypted value's storage format, which should help for consistency.

However, even in this form, there are some things that are done purely by convention: for example, for the AES GCM values, the `<b64-encoded-ciphertext>` is actually `b64_encode( IV bytes + ciphertext bytes )`, which is simply a convention chosen by `encrypted-config-value`. Implementations in different languages or formats must simply know that this is the convention and follow it.

Backcompat
-------------
In the new format, encrypted values will always have at least 3 colon-separated components within the braces (`enc`, `<key-alg-name>`, `<b64-encoded-ciphertext>`), whereas in the previous format there were only 2 (`enc`, `<b64-encoded-ciphertext>`).

If an encrypted value only has 2 components, it is assumed to be produced by an old version, and will be read as such using whatever key is currently in use. For values encrypted with AES keys, this means reading it as an AES value encrypted in GCM mode with a 32-byte IV and 16-byte tag length.

Upgrade Story
---------------
The Java bundle can be updated by a project at any time. The bundle will continue to read old keys and values fine, and any values produced in the new format will ostensibly be meant for the service that is producing the values, so it will be able to read those values.

Infrastructure services that may consume values produced by different `encrypted-config-value` products should be able to read both the new and old format as soon as there is a product that starts producing the new format. However, it should only produce output in the old format if such output may be consumed by old versions of the library. Any values that will be used only by the service or by a specific service that supports the newer version can use the new format.

nmiyake commented:
@markelliot @bavardage 

I know there's a lot going on here, but tried to summarize current state of things and offer a concrete proposal. Happy to chat/use another medium to iterate as well if you think that would be more effective.

nmiyake commented:
I know we already have a decent number of eyes on this, but tagging @uschi2000 since we chatted about this a bit in person earlier today

sfackler commented:
One minor modification I'd make is to change `${enc:AES:<IV+ciphertext+tag>:GCM:12:16}` to `${enc:AES:<ciphertext>:GCM:<IV>:<tag>}`. It simplifies a bit since there's no need to split the ciphertext block up manually when decrypting and those values aren't really part of the ciphertext per se.

Should it be `AES-256` rather than `AES` for the key prefix, for roughly the same reasons that we split `RSA-PUB` and `RSA-PRIV` up?

Seems good otherwise, in particular making the prefixes of the keys and values are identical.

nmiyake commented:
In most cases I don't think that the size of the key changes the manner in which it's deserialized, so don't feel as strongly about this (at least in Go, deserializing `AES-256` versus `AES-128` would be identical -- just read the bytes).

Also, if we encode the key size, then we should probably be consistent and have `RSA-PUB-2048` etc.

sfackler commented:
Ah right, that makes sense. Let's leave it as just `AES` then.

markelliot commented:
it makes more sense to me that we'd base64 encode a json object representing the details than to continue colon-separating the fields. The readability is neither helped nor harmed with this, but it makes it less likely that someone believes they can simply change one of these colon delimited things and have stuff work.

sfackler commented:
That seems okay to me - it will mean that we'll have to double-encode the ciphertext, IV, and tag though. I don't think that overall string length is something that we need to be particularly concerned about though. I'm imagining something like

```json
{
   "type": "AES",
   "mode": "GCM",
   "ct": "<base64 ciphertext>",
   "iv": "<base64 IV>",
   "tag": "<base64 tag>",
}
```
and
```json
{
    "type": "RSA-PUB",
    "ct": "<base64 ciphertext>",
}
```

right?

markelliot commented:
Yes, that seems fine to me.

nmiyake commented:
@sfackler your last comment outlines the objects in JSON -- is your proposal that we move to that representation instead?

If we do that, it's going to be a pretty large change/departure. There's a lot of code right now that depends on being able to identify encrypted values by searching with a regxp that captures `${enc:<.....>}` and then being able to interpret the string.

If we move to JSON, then suddenly it becomes much harder to handle cases like having encrypted values in the middle of a string. It also requires all clients of the library to be able to parse JSON and have the object definitions. I think this would merit a much larger discussion.

sfackler commented:
The end format would be the same - it's just that instead of the base64 portion being a raw ciphertext/other stuff, it's that JSON format. The final string for the AES example would be `${enc:eyJ0eXBlIjoiQUVTIiwibW9kZSI6IkdDTSIsImN0IjoiPGJhc2U2NCBjaXBoZXJ0ZXh0PiIsIml2IjoiPGJhc2U2NCBJVj4iLCJ0YWciOiI8YmFzZTY0IHRhZz4ifQ==}`.

nmiyake commented:
Ah OK yup that makes sense -- as long as final output format is stable/still a plain string, don't have any objections.

In that case, you're basically just talking about the Java object representation (rather than JSON), right? In that case, do we need to base64-encode as well? Encoding only really makes sense for moving things around in the serialized form. If we've read that form and are instantiating in memory, seems like it would make more sense to have:

```
{
   "type": "AES",
   "mode": "GCM",
   "ct": "ciphertext bytes",
   "iv": "IV bytes",
   "tag": "tag bytes",
}
```

Then, when we create the string format, we just do `base64(IV + ciphertext + tag)`. When we're reading out, we have to decode this full string anyway, so at that point might as well just store the constituent parts directly rather than re-encoding them just for storage.

nmiyake commented:
Also, looking at our current Java and Go code, it seems like the tag for GCM is almost always abstracted away in the algorithm itself -- it's not explicitly returned separately in any of the calls. Does this mean we just have to parse out the final bytes ourselves? A bit confused about what this would look like.

sfackler commented:
I was talking about the JSON - the point here is to I believe remove some of the implicit-ness around e.g. the ordering of the various bits and pieces of the encrypted value.

In many implementations (Go, Java, BoringSSL), the authentication tag is appended to the ciphertext, while in others (OpenSSL) it's handled separately.

nmiyake commented:
OK on-board with the tag part.

Still confused a bit about JSON... So you want to define a JSON format for this so that it's more explicit, but the serialization/deserialization will still be to/from the string form. In that case, what is defining the format as JSON actually providing?

nmiyake commented:
Unless the idea is that we define this JSON format and then just serialize to/from strings in config for back-compat, but use JSON form as a standard for other things. However, what would those "other" consumers be in that case?"

markelliot commented:
The point is just to clean up the semantics; you have to know how to parse the encoded string one way or another. Today it's under-specified, and so when we want to change the IV length it's not possible to have both old-and-new stuff parsed explicitly.

nmiyake commented:
Sorry for all the confusion -- chatted with Steven in person and cleared this up. I missed that the proposal is to change it such that new format is:

`${enc:<b64-encoded-json-object>}`, where the "JSON object" is something of the form above. This means that the library will now have JSON definitions for the encryption formats, and reading the `enc` value involves decoding the b64, then unmarshalling that into the proper JSON struct (will be polymorphic decryption, but `type` tag should make resolution fairly simple).

That was the part that I was missing -- now that I understand that, this all makes sense and I'm in agreement :)

nmiyake commented:
@sfackler one more follow-up -- in the current scheme, how (if at all) is it possible to distinguish between old `${enc:...}` strings and new ones? Is the plan to just decode the B64, check whether or not it's valid JSON and then decode as new format if it is, or or should it be encoded in the string itself (something like `${enc:v2:....}`)?

sfackler commented:
I was planning on just blindly trying the JSON codepath, catching any exception that's thrown, trying the fallback path, and if an exception's thrown there rethrow the JSON-codepath exception.

nmiyake commented:
OK -- I generally don't love fail-over deserialization, but JSON vs. non-JSON is pretty straightforward and if we're trying to move to this eventually for everything anyway not going to make a fuss about it.

sfackler committed:
Reform encrypted value storage

The base64'd contents are now a JSON blob with algorithm-specific
information, and the IV size for AES is now the proper value of 92 bits.
There are codepaths to be able to decrypt the old format, but all values
produced are in the new format.

sfackler commented:
Updated.

For a 43 byte plaintext, the old AES encoded format was 128 byes and the new AES encoded format is 220 bytes.

nmiyake commented on encrypted-config-value/src/main/java/com/palantir/config/crypto/algorithm/AesAlgorithm.java:
```diff
@@ -40,7 +44,7 @@
 
     private static final int GCM_AUTH_TAG_LENGTH = 128;
-    private static final int IV_LENGTH = 256 / Byte.SIZE;
+    private static final int IV_LENGTH = 96 / Byte.SIZE;
```
Can we have `TAG_LENGTH` and `IV_LENGTH` use the same units for consistency? Since these are private variables we should be able to change this without breaking anything

nmiyake replied:
Also, bigger question -- should we keep these as constants, but make them configurable but just have a default constructor/factory method that initializes with the default?

sfackler replied:
Yep, will fix up the units.

I am pretty against making these configurable at all. There is AFAIK zero reason to not use 12 byte IVs and 16 byte tags.

nmiyake replied:
OK yup, agree that there's not a compelling reason to make this configurable for new ciphers/values created by the library.

nmiyake commented on encrypted-config-value/src/main/java/com/palantir/config/crypto/value/EncryptedValue.java:
```diff
@@ -0,0 +1,115 @@
+     * @return the decrypted value
+     */
+    @JsonIgnore
+    public final String getDecryptedValue() {
```
Should we make this a static method/move outside of the class? I know we need to provide the raw functionality for back-compat, but this is somewhat external so feels strange to bake it in as behavior of this new type.

sfackler replied:
Yeah, this method in particular feels kind of weird and overly magical. It also means we're reading the key off disk and parsing it for every encrypted value in a config.

nmiyake commented on encrypted-config-value-bundle-dropwizard1/src/main/java/com/palantir/config/crypto/EncryptConfigValueCommand.java:
```diff
@@ -60,6 +61,6 @@ public void run(Bootstrap<?> bootstrap, Namespace namespace) throws Exception {
 
         // print the resulting encrypted value to the console
-        System.out.println(encryptedValue);
+        System.out.println(encryptedValue.serialize());
```
Would it make sense to define the `toString` to return the serialized representation so that these calls can stay as-is?

sfackler replied:
Checkstyle's design-for-extension lint and Immutables's setup become kind of unhappy with overriding toString at the `EncryptedValue` level, but I think I can hammer on things enough to make it work.

nmiyake replied:
OK -- I don't feel super strongly about this (requiring a call to `serialize()` is more semantically consistent), but was just a thought.

nmiyake commented on encrypted-config-value/src/main/java/com/palantir/config/crypto/value/AesEncryptedValue.java:
```diff
@@ -0,0 +1,63 @@
+@JsonDeserialize(as = ImmutableAesEncryptedValue.class)
+@JsonSerialize(as = ImmutableAesEncryptedValue.class)
+@JsonTypeName("AES")
+public abstract class AesEncryptedValue extends EncryptedValue {
```
Should this be `AesGcmEncryptedValue` instead? If we have a non-GCM AES value, then I think it would need its own class since fields like `iv` and `tag` may not be applicable (and there may be other fields as well).

In all likelihood we probably won't add another AES cipher type so probably fine, but just flagging.

sfackler replied:
Yeah I'd be okay merging the mode field into the type itself.

nmiyake reviewed (commented):
Some minor questions/feedback, but core of the change looks good to me.

nmiyake commented on encrypted-config-value/src/main/java/com/palantir/config/crypto/value/RsaEncryptedValue.java:
```diff
@@ -0,0 +1,40 @@
+@JsonSerialize(as = ImmutableRsaEncryptedValue.class)
+@JsonDeserialize(as = ImmutableRsaEncryptedValue.class)
+@JsonTypeName("RSA")
+public abstract class RsaEncryptedValue extends EncryptedValue {
```
Happy to do this as a separate/follow-up PR, but just as `AES-GCM` has specific configuration values, `RSA-OAEP-MGF1` has configuration values as well -- specifically, the hash function for `OAEP` and `MGF1` should both be separately configurable as either `SHA-1` or `SHA-256`. The legacy values use `SHA-256` for `OAEP` and `SHA-1` for `MGF1`, but we may also want to change the "new" values to be the more standard `SHA-256` for both (in which case we will also need to update the algorithm class).

See https://github.com/palantir/encrypted-config-value/issues/58#issuecomment-294089060 for details.

nmiyake commented on encrypted-config-value/src/main/java/com/palantir/config/crypto/value/AesEncryptedValue.java:
```diff
@@ -0,0 +1,63 @@
+
+    public abstract Mode getMode();
+
+    public abstract byte[] getIv();
```
One more question -- what controls the JSON serialization format of this object right now? Since byte arrays don't have a direct representation in JSON, assume that we need to define a serializer that turns the byte array fields into base64-encoded strings. That would mean that raw JSON would be:

```
{
   "mode": ...,
   "iv:"<b64-encoded-iv>",
   "ciphertext":"<b64-encoded-ciphertext>",
...
}
```

Then the "serialized" representation is the B64-encoded version of this JSON string.

Apologies if you're already doing this somewhere, but looked around and didn't see any code that seems to do this in the PR (I see that there's a call to `return MAPPER.writeValueAsBytes(aesEncryptedValue);`, but can't tell how this call will handle the `byte[]` fields).

sfackler replied:
Jackson will automatically use base64 for byte arrays, yeah.

nmiyake replied:
👍

nmiyake commented:
Implemented a PR that matches this implementation on the Go side, and I think this works out really well -- looking forward for the Java side to merge so that we can get this out!

FYI, here are some samples from that side (AES should match this PR, RSA will require follow-up PR):

AES
----
Plaintext: `plaintext`
Serialized key: `AES:LICx0yKzQm5a6IE13aJ3xOsRv+8AujqHocTFI4yk4Jw=`
JSON of `EncryptedValue`:
```json
{
  "type":"AES",
  "mode":"GCM",
  "ciphertext":"M94kIyoa5+2Z",
  "iv":"uAGqRlP9wizpdB0z",
  "tag":"ACSuzDwTULomsjxpFMkYKA=="
}
```

Serialized `EncryptedValue`: `${enc:eyJ0eXBlIjoiQUVTIiwibW9kZSI6IkdDTSIsImNpcGhlcnRleHQiOiJNOTRrSXlvYTUrMloiLCJpdiI6InVBR3FSbFA5d2l6cGRCMHoiLCJ0YWciOiJBQ1N1ekR3VFVMb21zanhwRk1rWUtBPT0ifQ==}`

RSA
----
Plaintext: `plaintext`
Serialized private key: `RSA-PRIV:MIIEvAIBADALBgkqhkiG9w0BAQEEggSoMIIEpAIBAAKCAQEAtKj+bpwUCq22ABjeJLBje+mD5XWmUAc8K2NbEGNGFaWGVAE1h/2Pjgxmj+LR4Bgt3OleYOnfV99ToqMNgB+HnNOJCg5LkHfq+WD6tRwhxFQMCmt73k9i8fgg7OCb1yTWo6pLCBIVWeisO0j0b1CYeIHebRemkx+8AK0ebsv4tdrIwAlb4jJTSz2rKZpEw7rLcGr8dFOYP5pg/jLJneittODD/uJj+1lpOze/AUT3bcuF6Ku0Oh4zNIvPcmm72bbr7+61lFOJB1IbDg1ahklE9m439/OOi3OOTdqq/HOu0k/dThrvovV1eedoL6UQz6RdijHNUt3iZqiues/Mq5dLSwIDAQABAoIBAQCEQyTi/cl+d+bC83HPEoQC99bkatmzxVg7u6WzvbpVprVNUwVJ5kzvBg0gUkKs+Ya6MPAzq4Uj5BBrBUyg/HRgUE4H2qdfwSt6H5HsfggKoC2gg0hQXXZnB+2y/k2ZmRK7B7We1v5isIFHdgXeaPb3YrzgyWveUmFlbVjWbOZM3AAJ0FczP2b3DErFS/iMyzdjCY9xwwXhQediMASj24c44/VLsaRCFesPXHoAXCvLLlPmNhfaw6ZVtHblg0QlFNftOUlIXC+s9yIN2ec38C10VR/yfGqVSYz+owXqNKRpMfsqNe1jWnl3+BVaqO53vsXzkYU8n8/vHdRSRZOiKpwhAoGBAMVutRUefOcApu5iEpHK+7Jte0o1kNFIwCXqiujjZcU/DKjDj2yK90ioza7Ntp7EHI9MUgCknyyiMlI/1VtHl3KiNfi0FQ646/AxOgzfrmUZTTyUgq02ToxFnAr1XYBzwwAPHKM4p2nJrf+G/7FpXhCMhK4qwGfMJ4C+i0pCoiJ/AoGBAOpAktM4SZGmBdtPyRpp0Z8tkrHoRNwn1YK+VS7XfkKmeMrsPEev7cjesaNJnMBjtlpGrAzVJC/ycEz5lZW7gBA5i/hDOLGegLjuu1SOTKXU4IFw5+vjTe4ecMFLLRE/rTeWMR3RfslzTiV66zLKZ9zuhq4YccGi3IFKKXVp+Fk1AoGAN86kVxToH2/6v7VvJFDpNrVlvUNI7S+QSOd0XoIwuUGqNWYZ+4eIgLxeb4PslBJBNGxRXacq6zXp3X/3sjaZY6jgcq2Mqj2xS5LOoubzZ9ZwE6izC30nVNU0V5Cl3nJac4DSCn0wLWH50hn52s867JibxJOHEZAOtoCl5NbS98cCgYAWoCoOUK96a+jA6BHqhTIEB+jVWjPcd9R9jli374R4d4/POcYQvoNfFXNe7CtBwd/JFG5lxuh54RbLuIekMLoL1yMX1ZZSQZb5RcW+QwhQNCGDHx6ngAr05ufJI7O0qMvYRJ9129g9KO/xWtAA1d/2TOuhQScrpslZi4o5lwSvyQKBgQCw/nLpPPlBGeA6jA0yZOuMPDZMGStLOAsGMmhV6LnBBllE475qQRPD/1xgcoWU7+u9H6sJNBR5p/WJq58IZFHzVCFVEBijLbNXDKOF9nDaczzXID5pM2Pspoz7JPpZkIFk0D2IR73M2RfoWNxYPRJCImDaL7HOXND6SNA+p6kkMg==`
JSON of `EncryptedValue`:
```json
{
  "type":"RSA",
  "mode":"OAEP",
  "ciphertext":"EwGDCUqzoC3QyQxOuxpNlsasZWNBnE7GvmWEYsJRmlls8GK831rCC6HfFyxfZhsSqjIZJfu19NtdDWNRYVyV+zz8OMvNfN6UxZqE6ptQvGTGmGABOB+kEdhMxVEGBwMoXJxtfX9Ji4dxKtzYfXDliXuu895pz9P7iv6oFGd5SmONT0YskkjbRY0hTXEeztioEm00sWjO2Jrguu05FyJHfFJH5+Op4ldVdFp5ejV+lQ8EL5EAZ2h7+lqV9FjtzlAukqmJx8Ep0hIlhOR236wIqd0dzAW6g862lCKoShoGoKN2ttyjU6MDQQj4kHA4WBm0NvoWDFG7ZwOZ3l20Vv7GVQ==",
  "oaep-alg":"SHA-256",
  "mdf1-alg":"SHA-256"
}
```

Serialized `EncryptedValue`: `${enc:eyJ0eXBlIjoiUlNBIiwibW9kZSI6Ik9BRVAiLCJjaXBoZXJ0ZXh0IjoiRXdHRENVcXpvQzNReVF4T3V4cE5sc2FzWldOQm5FN0d2bVdFWXNKUm1sbHM4R0s4MzFyQ0M2SGZGeXhmWmhzU3FqSVpKZnUxOU50ZERXTlJZVnlWK3p6OE9Ndk5mTjZVeFpxRTZwdFF2R1RHbUdBQk9CK2tFZGhNeFZFR0J3TW9YSnh0Zlg5Smk0ZHhLdHpZZlhEbGlYdXU4OTVwejlQN2l2Nm9GR2Q1U21PTlQwWXNra2piUlkwaFRYRWV6dGlvRW0wMHNXak8ySnJndXUwNUZ5SkhmRkpINStPcDRsZFZkRnA1ZWpWK2xROEVMNUVBWjJoNytscVY5Rmp0emxBdWtxbUp4OEVwMGhJbGhPUjIzNndJcWQwZHpBVzZnODYybENLb1Nob0dvS04ydHR5alU2TURRUWo0a0hBNFdCbTBOdm9XREZHN1p3T1ozbDIwVnY3R1ZRPT0iLCJvYWVwLWFsZyI6IlNIQS0yNTYiLCJtZGYxLWFsZyI6IlNIQS0yNTYifQ==}`

sfackler commented:
Cool - is this ready to land then?

markelliot commented:
yep, this seems fine.

sfackler commented:
Cool - I don't have write access. Could someone who does merge?
